File: frontend/static/js/binary-search-tree.js
// Binary Search Tree Visualization
// Author: Aryan Pravin Sahu

class TreeNode {
    constructor(value) {
        this.value = value;
        this.left = null;
        this.right = null;
        this.x = 0;
        this.y = 0;
        this.element = null;
    }
}

class BSTVisualizer {
    constructor() {
        this.root = null;
        this.isAnimating = false;
        this.nodeSize = 50;
        this.levelHeight = 80;
        this.nodeSpacing = 60;
        
        this.initializeElements();
        this.setupEventListeners();
        this.updateVisualization();
    }

    initializeElements() {
        // Control elements
        this.valueInput = document.getElementById('valueInput');
        this.bulkInput = document.getElementById('bulkInput');
        
        // Button elements
        this.insertBtn = document.getElementById('insertBtn');
        this.searchBtn = document.getElementById('searchBtn');
        this.deleteBtn = document.getElementById('deleteBtn');
        this.clearBtn = document.getElementById('clearBtn');
        this.inorderBtn = document.getElementById('inorderBtn');
        this.preorderBtn = document.getElementById('preorderBtn');
        this.postorderBtn = document.getElementById('postorderBtn');
        this.bulkInsertBtn = document.getElementById('bulkInsertBtn');
        this.demoBtn = document.getElementById('demoBtn');
        
        // Visualization elements
        this.treeVisualization = document.getElementById('treeVisualization');
        this.emptyMessage = document.getElementById('emptyMessage');
        this.stepInfo = document.getElementById('stepInfo');
        this.traversalResult = document.getElementById('traversalResult');
        this.traversalTitle = document.getElementById('traversalTitle');
        this.traversalValues = document.getElementById('traversalValues');
        
        // Array visualization elements
        this.arrayVisualization = document.getElementById('arrayVisualization');
        this.arrayContainer = document.getElementById('arrayContainer');
        
        // Current traversal state
        this.currentTraversal = 'inorder';
        this.traversalArray = [];
        
        // Info elements
        this.treeSize = document.getElementById('treeSize');
        this.treeHeight = document.getElementById('treeHeight');
        this.rootValue = document.getElementById('rootValue');
        this.lastOperation = document.getElementById('lastOperation');
    }

    setupEventListeners() {
        this.insertBtn.addEventListener('click', () => this.insert());
        this.searchBtn.addEventListener('click', () => this.search());
        this.deleteBtn.addEventListener('click', () => this.delete());
        this.clearBtn.addEventListener('click', () => this.clear());
        this.inorderBtn.addEventListener('click', () => this.traverseInOrder());
        this.preorderBtn.addEventListener('click', () => this.traversePreOrder());
        this.postorderBtn.addEventListener('click', () => this.traversePostOrder());
        this.bulkInsertBtn.addEventListener('click', () => this.bulkInsert());
        this.demoBtn.addEventListener('click', () => this.runDemo());

        // Allow Enter key to insert
        this.valueInput.addEventListener('keypress', (e) => {
            if (e.key === 'Enter' && !this.isAnimating) {
                this.insert();
            }
        });

        // Traversal button event listeners
        document.querySelectorAll('.traversal-btn').forEach(btn => {
            btn.addEventListener('click', (e) => {
                const traversalType = e.target.dataset.type;
                this.setTraversalType(traversalType);
            });
        });
    }

    async insert(value = null) {
        if (this.isAnimating) return;

        const inputValue = value !== null ? value : parseInt(this.valueInput.value);
        if (isNaN(inputValue)) {
            alert('Please enter a valid number');
            return;
        }

        this.isAnimating = true;
        this.disableControls();

        // Check if value already exists
        if (this.findNode(this.root, inputValue)) {
            alert(`Value ${inputValue} already exists in the tree`);
            this.enableControls();
            this.isAnimating = false;
            return;
        }

        // Insert the node
        this.root = await this.insertNode(this.root, inputValue);
        
        // Update visualization
        this.calculatePositions();
        this.renderTree();
        this.updateVisualization();
        
        this.updateStepInfo(`Inserted ${inputValue} into the BST`, 
                           `Node ${inputValue} has been added to the tree following BST property.`);
        
        // Clear input
        if (value === null) {
            this.valueInput.value = '';
        }

        this.enableControls();
        this.isAnimating = false;
    }

    async insertNode(node, value) {
        if (node === null) {
            const newNode = new TreeNode(value);
            await this.animateNodeInsertion(newNode);
            return newNode;
        }

        if (value < node.value) {
            node.left = await this.insertNode(node.left, value);
        } else if (value > node.value) {
            node.right = await this.insertNode(node.right, value);
        }

        return node;
    }

    async animateNodeInsertion(node) {
        // Create visual element
        const element = this.createNodeElement(node.value);
        element.classList.add('inserting');
        this.treeVisualization.appendChild(element);
        node.element = element;

        await this.sleep(400);
        element.classList.remove('inserting');
    }

    async search(value = null) {
        if (this.isAnimating) return;

        const searchValue = value !== null ? value : parseInt(this.valueInput.value);
        if (isNaN(searchValue)) {
            alert('Please enter a valid number to search');
            return;
        }

        if (!this.root) {
            alert('Tree is empty');
            return;
        }

        this.isAnimating = true;
        this.disableControls();

        // Clear previous search highlighting
        this.clearHighlighting();

        const result = await this.searchNode(this.root, searchValue);
        
        if (result) {
            this.updateStepInfo(`Found ${searchValue} in the BST`, 
                               `Value ${searchValue} exists in the tree and has been highlighted.`);
        } else {
            this.updateStepInfo(`${searchValue} not found in the BST`, 
                               `Value ${searchValue} does not exist in the tree.`);
        }

        this.enableControls();
        this.isAnimating = false;
    }

    async searchNode(node, value) {
        if (node === null) {
            return false;
        }

        // Highlight current node
        if (node.element) {
            node.element.classList.add('searching');
            await this.sleep(600);
        }

        if (value === node.value) {
            // Found the value
            if (node.element) {
                node.element.classList.remove('searching');
                node.element.classList.add('found');
            }
            return true;
        } else if (value < node.value) {
            // Remove searching class and search left
            if (node.element) {
                node.element.classList.remove('searching');
            }
            return await this.searchNode(node.left, value);
        } else {
            // Remove searching class and search right
            if (node.element) {
                node.element.classList.remove('searching');
            }
            return await this.searchNode(node.right, value);
        }
    }

    async delete(value = null) {
        if (this.isAnimating) return;

        const deleteValue = value !== null ? value : parseInt(this.valueInput.value);
        if (isNaN(deleteValue)) {
            alert('Please enter a valid number to delete');
            return;
        }

        if (!this.root) {
            alert('Tree is empty');
            return;
        }

        // Check if value exists
        if (!this.findNode(this.root, deleteValue)) {
            alert(`Value ${deleteValue} not found in the tree`);
            return;
        }

        this.isAnimating = true;
        this.disableControls();

        // Delete the node
        this.root = await this.deleteNode(this.root, deleteValue);
        
        // Update visualization
        this.calculatePositions();
        this.renderTree();
        this.updateVisualization();
        
        this.updateStepInfo(`Deleted ${deleteValue} from the BST`, 
                           `Node ${deleteValue} has been removed from the tree.`);

        this.enableControls();
        this.isAnimating = false;
    }

    async deleteNode(node, value) {
        if (node === null) {
            return null;
        }

        if (value < node.value) {
            node.left = await this.deleteNode(node.left, value);
        } else if (value > node.value) {
            node.right = await this.deleteNode(node.right, value);
        } else {
            // Node to be deleted found
            if (node.element) {
                node.element.classList.add('deleting');
                await this.sleep(400);
                node.element.remove();
            }

            // Case 1: No children
            if (node.left === null && node.right === null) {
                return null;
            }
            // Case 2: One child
            else if (node.left === null) {
                return node.right;
            } else if (node.right === null) {
                return node.left;
            }
            // Case 3: Two children
            else {
                // Find inorder successor (smallest in right subtree)
                const successor = this.findMin(node.right);
                node.value = successor.value;
                node.right = await this.deleteNode(node.right, successor.value);
            }
        }

        return node;
    }

    findMin(node) {
        while (node.left !== null) {
            node = node.left;
        }
        return node;
    }

    findNode(node, value) {
        if (node === null) {
            return null;
        }
        if (value === node.value) {
            return node;
        } else if (value < node.value) {
            return this.findNode(node.left, value);
        } else {
            return this.findNode(node.right, value);
        }
    }

    async traverseInOrder() {
        if (!this.root) {
            alert('Tree is empty');
            return;
        }

        const result = [];
        await this.inOrderTraversal(this.root, result);
        this.showTraversalResult('In-order Traversal (Left → Root → Right)', result);
    }

    async traversePreOrder() {
        if (!this.root) {
            alert('Tree is empty');
            return;
        }

        const result = [];
        await this.preOrderTraversal(this.root, result);
        this.showTraversalResult('Pre-order Traversal (Root → Left → Right)', result);
    }

    async traversePostOrder() {
        if (!this.root) {
            alert('Tree is empty');
            return;
        }

        const result = [];
        await this.postOrderTraversal(this.root, result);
        this.showTraversalResult('Post-order Traversal (Left → Right → Root)', result);
    }

    async inOrderTraversal(node, result) {
        if (node !== null) {
            await this.inOrderTraversal(node.left, result);
            
            // Highlight current node
            if (node.element) {
                node.element.classList.add('searching');
                await this.sleep(500);
                node.element.classList.remove('searching');
            }
            
            result.push(node.value);
            await this.inOrderTraversal(node.right, result);
        }
    }

    async preOrderTraversal(node, result) {
        if (node !== null) {
            // Highlight current node
            if (node.element) {
                node.element.classList.add('searching');
                await this.sleep(500);
                node.element.classList.remove('searching');
            }
            
            result.push(node.value);
            await this.preOrderTraversal(node.left, result);
            await this.preOrderTraversal(node.right, result);
        }
    }

    async postOrderTraversal(node, result) {
        if (node !== null) {
            await this.postOrderTraversal(node.left, result);
            await this.postOrderTraversal(node.right, result);
            
            // Highlight current node
            if (node.element) {
                node.element.classList.add('searching');
                await this.sleep(500);
                node.element.classList.remove('searching');
            }
            
            result.push(node.value);
        }
    }

    showTraversalResult(title, values) {
        this.traversalTitle.textContent = title;
        this.traversalValues.textContent = values.join(' → ');
        this.traversalResult.style.display = 'block';
        
        setTimeout(() => {
            this.traversalResult.style.display = 'none';
        }, 5000);
    }

    async bulkInsert() {
        if (this.isAnimating) return;

        const input = this.bulkInput.value.trim();
        if (!input) {
            alert('Please enter comma-separated values');
            return;
        }

        const values = input.split(',')
            .map(val => parseInt(val.trim()))
            .filter(val => !isNaN(val));
        
        if (values.length === 0) {
            alert('Please enter valid numbers');
            return;
        }

        for (const value of values) {
            await this.insert(value);
            await this.sleep(300);
        }
    }

    async runDemo() {
        if (this.isAnimating) return;

        // Clear tree first
        this.clear();
        await this.sleep(500);

        const demoValues = [50, 30, 70, 20, 40, 60, 80, 10, 25, 35, 45];
        
        this.updateStepInfo('Running BST Demo', 'Building a sample binary search tree...');

        // Insert demo values
        for (const value of demoValues) {
            await this.insert(value);
            await this.sleep(400);
        }

        await this.sleep(1000);

        // Demonstrate search
        this.updateStepInfo('Demo: Searching for 25', 'Demonstrating search operation...');
        await this.search(25);
        
        await this.sleep(2000);

        // Demonstrate traversal
        this.updateStepInfo('Demo: In-order Traversal', 'Showing sorted order of values...');
        await this.traverseInOrder();

        this.updateStepInfo('Demo completed', 'BST demonstration finished. Try your own operations!');
    }

    clear() {
        if (this.isAnimating) return;

        this.root = null;
        
        // Remove all visual elements
        const nodes = this.treeVisualization.querySelectorAll('.tree-node');
        const edges = this.treeVisualization.querySelectorAll('.tree-edge');
        
        nodes.forEach(node => node.remove());
        edges.forEach(edge => edge.remove());

        this.emptyMessage.style.display = 'flex';
        this.traversalResult.style.display = 'none';
        
        this.updateVisualization();
        this.updateStepInfo('Tree cleared', 'All nodes have been removed from the BST.');
    }

    calculatePositions() {
        if (!this.root) return;

        const width = this.treeVisualization.offsetWidth;
        const startX = width / 2;
        const startY = 50;

        this.assignPositions(this.root, startX, startY, width / 4);
    }

    assignPositions(node, x, y, offset) {
        if (node === null) return;

        node.x = x;
        node.y = y;

        if (node.left) {
            this.assignPositions(node.left, x - offset, y + this.levelHeight, offset / 2);
        }
        if (node.right) {
            this.assignPositions(node.right, x + offset, y + this.levelHeight, offset / 2);
        }
    }

    renderTree() {
        // Clear existing visual elements
        const nodes = this.treeVisualization.querySelectorAll('.tree-node');
        const edges = this.treeVisualization.querySelectorAll('.tree-edge');
        
        nodes.forEach(node => node.remove());
        edges.forEach(edge => edge.remove());

        if (!this.root) {
            this.emptyMessage.style.display = 'flex';
            return;
        }

        this.emptyMessage.style.display = 'none';

        // Render edges first (so they appear behind nodes)
        this.renderEdges(this.root);
        
        // Render nodes
        this.renderNodes(this.root);
    }

    renderNodes(node) {
        if (node === null) return;

        const element = this.createNodeElement(node.value);
        element.style.left = (node.x - this.nodeSize / 2) + 'px';
        element.style.top = node.y + 'px';
        
        if (node === this.root) {
            element.classList.add('root');
        }

        this.treeVisualization.appendChild(element);
        node.element = element;

        this.renderNodes(node.left);
        this.renderNodes(node.right);
    }

    renderEdges(node) {
        if (node === null) return;

        if (node.left) {
            this.createEdge(node, node.left);
            this.renderEdges(node.left);
        }
        if (node.right) {
            this.createEdge(node, node.right);
            this.renderEdges(node.right);
        }
    }

    createEdge(parent, child) {
        const edge = document.createElement('div');
        edge.className = 'tree-edge';

        const dx = child.x - parent.x;
        const dy = child.y - parent.y;
        const length = Math.sqrt(dx * dx + dy * dy);
        const angle = Math.atan2(dy, dx) * 180 / Math.PI;

        edge.style.left = parent.x + 'px';
        edge.style.top = (parent.y + this.nodeSize / 2) + 'px';
        edge.style.width = length + 'px';
        edge.style.transform = `rotate(${angle}deg)`;

        this.treeVisualization.appendChild(edge);
    }

    createNodeElement(value) {
        const element = document.createElement('div');
        element.className = 'tree-node';
        element.textContent = value;
        element.dataset.value = value;
        return element;
    }

    clearHighlighting() {
        const nodes = this.treeVisualization.querySelectorAll('.tree-node');
        nodes.forEach(node => {
            node.classList.remove('searching', 'found');
        });
    }

    getTreeSize(node = this.root) {
        if (node === null) return 0;
        return 1 + this.getTreeSize(node.left) + this.getTreeSize(node.right);
    }

    getTreeHeight(node = this.root) {
        if (node === null) return 0;
        return 1 + Math.max(this.getTreeHeight(node.left), this.getTreeHeight(node.right));
    }

    updateVisualization() {
        // Update info display
        this.treeSize.textContent = this.getTreeSize();
        this.treeHeight.textContent = this.getTreeHeight();
        this.rootValue.textContent = this.root ? this.root.value : 'None';
        
        // Update button states
        const hasNodes = this.root !== null;
        this.searchBtn.disabled = !hasNodes;
        this.deleteBtn.disabled = !hasNodes;
        this.clearBtn.disabled = !hasNodes;
        this.inorderBtn.disabled = !hasNodes;
        this.preorderBtn.disabled = !hasNodes;
        this.postorderBtn.disabled = !hasNodes;
        
        // Update array visualization
        this.updateArrayVisualization();
    }

    updateArrayVisualization() {
        // Get current traversal
        this.traversalArray = this.getTraversalArray(this.currentTraversal);
        
        // Clear previous array visualization
        this.arrayVisualization.innerHTML = '';
        
        if (this.traversalArray.length === 0) {
            const emptyDiv = document.createElement('div');
            emptyDiv.className = 'empty-array';
            emptyDiv.textContent = 'No elements to display';
            this.arrayVisualization.appendChild(emptyDiv);
            return;
        }
        
        // Create array elements
        this.traversalArray.forEach((value, index) => {
            const element = document.createElement('div');
            element.className = 'array-element';
            element.textContent = value;
            element.dataset.value = value;
            element.dataset.index = index;
            
            // Add click handler to highlight corresponding tree node
            element.addEventListener('click', () => {
                this.highlightNodeInTree(value);
                this.highlightArrayElement(index);
            });
            
            this.arrayVisualization.appendChild(element);
        });
    }

    getTraversalArray(type) {
        const result = [];
        
        switch (type) {
            case 'inorder':
                this.inorderTraversal(this.root, result);
                break;
            case 'preorder':
                this.preorderTraversal(this.root, result);
                break;
            case 'postorder':
                this.postorderTraversal(this.root, result);
                break;
        }
        
        return result;
    }

    inorderTraversal(node, result) {
        if (node) {
            this.inorderTraversal(node.left, result);
            result.push(node.value);
            this.inorderTraversal(node.right, result);
        }
    }

    preorderTraversal(node, result) {
        if (node) {
            result.push(node.value);
            this.preorderTraversal(node.left, result);
            this.preorderTraversal(node.right, result);
        }
    }

    postorderTraversal(node, result) {
        if (node) {
            this.postorderTraversal(node.left, result);
            this.postorderTraversal(node.right, result);
            result.push(node.value);
        }
    }

    highlightNodeInTree(value) {
        // Remove previous highlights
        document.querySelectorAll('.tree-node').forEach(node => {
            node.classList.remove('highlighted');
        });
        
        // Find and highlight the node with the given value
        const nodeElement = document.querySelector(`[data-value="${value}"]`);
        if (nodeElement && nodeElement.classList.contains('tree-node')) {
            nodeElement.classList.add('highlighted');
            
            // Remove highlight after 2 seconds
            setTimeout(() => {
                nodeElement.classList.remove('highlighted');
            }, 2000);
        }
    }

    highlightArrayElement(index) {
        // Remove previous highlights
        document.querySelectorAll('.array-element').forEach(element => {
            element.classList.remove('highlighted');
        });
        
        // Highlight selected element
        const element = document.querySelector(`[data-index="${index}"]`);
        if (element) {
            element.classList.add('highlighted');
            
            // Remove highlight after 2 seconds
            setTimeout(() => {
                element.classList.remove('highlighted');
            }, 2000);
        }
    }

    setTraversalType(type) {
        this.currentTraversal = type;
        
        // Update button states
        document.querySelectorAll('.traversal-btn').forEach(btn => {
            btn.classList.remove('active');
        });
        document.querySelector(`[data-type="${type}"]`).classList.add('active');
        
        // Update array visualization
        this.updateArrayVisualization();
        
        // Update step info
        const descriptions = {
            'inorder': 'In-Order Traversal: Left → Root → Right',
            'preorder': 'Pre-Order Traversal: Root → Left → Right', 
            'postorder': 'Post-Order Traversal: Left → Right → Root'
        };
        
        this.updateStepInfo(descriptions[type], 
                           `Showing ${type} traversal sequence. Click array elements to highlight corresponding tree nodes.`);
    }

    updateStepInfo(description, details) {
        const stepDescription = this.stepInfo.querySelector('.step-description');
        const stepDetails = this.stepInfo.querySelector('.step-details');
        
        stepDescription.textContent = description;
        stepDetails.textContent = details;
        
        this.lastOperation.textContent = description.split(' ')[0];
    }

    disableControls() {
        this.insertBtn.disabled = true;
        this.searchBtn.disabled = true;
        this.deleteBtn.disabled = true;
        this.clearBtn.disabled = true;
        this.inorderBtn.disabled = true;
        this.preorderBtn.disabled = true;
        this.postorderBtn.disabled = true;
        this.bulkInsertBtn.disabled = true;
        this.demoBtn.disabled = true;
    }

    enableControls() {
        this.insertBtn.disabled = false;
        this.bulkInsertBtn.disabled = false;
        this.demoBtn.disabled = false;
        this.updateVisualization(); // This will set the correct disabled states
    }

    sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}

// Initialize the visualizer when the page loads
document.addEventListener('DOMContentLoaded', function() {
    new BSTVisualizer();
});

File: backend/algorithms/trees/binary_search_tree.py
"""
Binary Search Tree Implementation with Tree Visualization
Author: Aryan Pravin Sahu
"""

class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None

class BinarySearchTree:
    def __init__(self):
        self.root = None
    
    def insert_steps(self, values):
        """Generate step-by-step BST insertion"""
        steps = []
        
        # Initial empty tree
        steps.append({
            'step': 0,
            'tree': self.serialize_tree(),
            'operation': 'initial',
            'current_node': None,
            'inserting_value': None,
            'path': [],
            'comparing': [],
            'description': 'Empty Binary Search Tree - ready for insertions'
        })
        
        step_count = 1
        
        for value in values:
            insertion_steps = self._insert_value_steps(value, step_count)
            steps.extend(insertion_steps)
            step_count += len(insertion_steps)
        
        return steps
    
    def _insert_value_steps(self, value, start_step):
        """Generate steps for inserting a single value"""
        steps = []
        step_count = start_step
        
        # Show value being inserted
        steps.append({
            'step': step_count,
            'tree': self.serialize_tree(),
            'operation': 'insert_start',
            'current_node': None,
            'inserting_value': value,
            'path': [],
            'comparing': [],
            'description': f'Inserting value {value} into BST'
        })
        step_count += 1
        
        if self.root is None:
            # First node becomes root
            self.root = TreeNode(value)
            steps.append({
                'step': step_count,
                'tree': self.serialize_tree(),
                'operation': 'insert_root',
                'current_node': value,
                'inserting_value': value,
                'path': [value],
                'comparing': [],
                'description': f'Tree is empty. {value} becomes the root node'
            })
            return steps
        
        # Traverse to find insertion point
        current = self.root
        path = []
        
        while True:
            path.append(current.val)
            
            # Show comparison
            steps.append({
                'step': step_count,
                'tree': self.serialize_tree(),
                'operation': 'compare',
                'current_node': current.val,
                'inserting_value': value,
                'path': path.copy(),
                'comparing': [current.val],
                'description': f'Comparing {value} with {current.val}'
            })
            step_count += 1
            
            if value < current.val:
                # Go left
                steps.append({
                    'step': step_count,
                    'tree': self.serialize_tree(),
                    'operation': 'go_left',
                    'current_node': current.val,
                    'inserting_value': value,
                    'path': path.copy(),
                    'comparing': [],
                    'description': f'{value} < {current.val}, go to left subtree'
                })
                step_count += 1
                
                if current.left is None:
                    # Insert here
                    current.left = TreeNode(value)
                    path.append(value)
                    steps.append({
                        'step': step_count,
                        'tree': self.serialize_tree(),
                        'operation': 'insert_complete',
                        'current_node': value,
                        'inserting_value': value,
                        'path': path.copy(),
                        'comparing': [],
                        'description': f'Inserted {value} as left child of {current.val}'
                    })
                    break
                else:
                    current = current.left
                    
            elif value > current.val:
                # Go right
                steps.append({
                    'step': step_count,
                    'tree': self.serialize_tree(),
                    'operation': 'go_right',
                    'current_node': current.val,
                    'inserting_value': value,
                    'path': path.copy(),
                    'comparing': [],
                    'description': f'{value} > {current.val}, go to right subtree'
                })
                step_count += 1
                
                if current.right is None:
                    # Insert here
                    current.right = TreeNode(value)
                    path.append(value)
                    steps.append({
                        'step': step_count,
                        'tree': self.serialize_tree(),
                        'operation': 'insert_complete',
                        'current_node': value,
                        'inserting_value': value,
                        'path': path.copy(),
                        'comparing': [],
                        'description': f'Inserted {value} as right child of {current.val}'
                    })
                    break
                else:
                    current = current.right
            else:
                # Duplicate value
                steps.append({
                    'step': step_count,
                    'tree': self.serialize_tree(),
                    'operation': 'duplicate',
                    'current_node': current.val,
                    'inserting_value': value,
                    'path': path.copy(),
                    'comparing': [current.val],
                    'description': f'Value {value} already exists in BST. Skipping insertion.'
                })
                break
        
        return steps
    
    def search_steps(self, target):
        """Generate step-by-step BST search"""
        steps = []
        
        # Initial state
        steps.append({
            'step': 0,
            'tree': self.serialize_tree(),
            'operation': 'search_start',
            'current_node': None,
            'target': target,
            'path': [],
            'comparing': [],
            'found': False,
            'description': f'Searching for {target} in BST'
        })
        
        if self.root is None:
            steps.append({
                'step': 1,
                'tree': self.serialize_tree(),
                'operation': 'not_found',
                'current_node': None,
                'target': target,
                'path': [],
                'comparing': [],
                'found': False,
                'description': f'Tree is empty. {target} not found.'
            })
            return steps
        
        step_count = 1
        current = self.root
        path = []
        
        while current is not None:
            path.append(current.val)
            
            # Show comparison
            steps.append({
                'step': step_count,
                'tree': self.serialize_tree(),
                'operation': 'compare',
                'current_node': current.val,
                'target': target,
                'path': path.copy(),
                'comparing': [current.val],
                'found': False,
                'description': f'Comparing target {target} with {current.val}'
            })
            step_count += 1
            
            if target == current.val:
                # Found
                steps.append({
                    'step': step_count,
                    'tree': self.serialize_tree(),
                    'operation': 'found',
                    'current_node': current.val,
                    'target': target,
                    'path': path.copy(),
                    'comparing': [current.val],
                    'found': True,
                    'description': f'Target {target} found!'
                })
                break
            elif target < current.val:
                # Go left
                steps.append({
                    'step': step_count,
                    'tree': self.serialize_tree(),
                    'operation': 'go_left',
                    'current_node': current.val,
                    'target': target,
                    'path': path.copy(),
                    'comparing': [],
                    'found': False,
                    'description': f'{target} < {current.val}, search left subtree'
                })
                step_count += 1
                current = current.left
            else:
                # Go right
                steps.append({
                    'step': step_count,
                    'tree': self.serialize_tree(),
                    'operation': 'go_right',
                    'current_node': current.val,
                    'target': target,
                    'path': path.copy(),
                    'comparing': [],
                    'found': False,
                    'description': f'{target} > {current.val}, search right subtree'
                })
                step_count += 1
                current = current.right
        
        # Not found
        if current is None:
            steps.append({
                'step': step_count,
                'tree': self.serialize_tree(),
                'operation': 'not_found',
                'current_node': None,
                'target': target,
                'path': path.copy(),
                'comparing': [],
                'found': False,
                'description': f'Target {target} not found in BST'
            })
        
        return steps
    
    def traversal_steps(self, traversal_type='inorder'):
        """Generate step-by-step tree traversal"""
        steps = []
        visited = []
        
        steps.append({
            'step': 0,
            'tree': self.serialize_tree(),
            'operation': f'{traversal_type}_start',
            'current_node': None,
            'visited': [],
            'stack': [],
            'result': [],
            'description': f'Starting {traversal_type} traversal'
        })
        
        if traversal_type == 'inorder':
            self._inorder_steps(self.root, steps, visited, [])
        elif traversal_type == 'preorder':
            self._preorder_steps(self.root, steps, visited, [])
        elif traversal_type == 'postorder':
            self._postorder_steps(self.root, steps, visited, [])
        
        return steps
    
    def _inorder_steps(self, node, steps, visited, result):
        """Inorder traversal with steps"""
        if node is not None:
            # Visit left
            self._inorder_steps(node.left, steps, visited, result)
            
            # Visit root
            visited.append(node.val)
            result.append(node.val)
            steps.append({
                'step': len(steps),
                'tree': self.serialize_tree(),
                'operation': 'visit',
                'current_node': node.val,
                'visited': visited.copy(),
                'stack': [],
                'result': result.copy(),
                'description': f'Visit node {node.val} (Inorder: Left → Root → Right)'
            })
            
            # Visit right
            self._inorder_steps(node.right, steps, visited, result)
    
    def serialize_tree(self):
        """Convert tree to serializable format for frontend"""
        if self.root is None:
            return None
        
        def serialize_node(node):
            if node is None:
                return None
            return {
                'val': node.val,
                'left': serialize_node(node.left),
                'right': serialize_node(node.right)
            }
        
        return serialize_node(self.root)

def get_sample_data():
    """Return sample values for BST demonstration"""
    return [50, 30, 70, 20, 40, 60, 80]

def get_sample_search_target():
    """Return sample search target"""
    return 40

# Test the implementation
if __name__ == "__main__":
    bst = BinarySearchTree()
    sample_values = get_sample_data()
    
    print(f"BST Insertion for values {sample_values}:")
    insertion_steps = bst.insert_steps(sample_values)
    print(f"Generated {len(insertion_steps)} insertion steps")
    
    print(f"\nBST Search for {get_sample_search_target()}:")
    search_steps = bst.search_steps(get_sample_search_target())
    print(f"Generated {len(search_steps)} search steps")
    
    print(f"\nBST Inorder Traversal:")
    traversal_steps = bst.traversal_steps('inorder')
    print(f"Generated {len(traversal_steps)} traversal steps")